Lib/concurrent/futures/process.py

```diff
@@ -72,15 +73,39 @@
 # workers to exit when their work queues are empty and then waits until the
 # threads/processes finish.
 
-_threads_queues = weakref.WeakKeyDictionary()
+_threads_sentinels = weakref.WeakKeyDictionary()
 _global_shutdown = False
 
+
+# This constant controls the maximal wakeup. If a job is submitted to the
+# Executor, it might take up to _POLL_TIMEOUT for the executor to notice and
+# start launching the job. This _POLL_TIMEOUT is to be cumulated with the
+# communication overhead.
+_POLL_TIMEOUT = .001
+
+
+class _Sentinel:
```

Comment:
In retrospect I find this class name confusing as it's very different from the worker process' `.sentinel` attribute. Maybe we should find a better name  e.g. `_ThreadNotifier`.

Reply:
Maybe putting `_ThreadWakeup` and `thread_wakeup` everywhere?

Reply:
Actually, it should be possible to rewrite this class to use an anonymous `os.pipe` to make it possible use it directly within the call to `multiprocessing.connection.wait` without relying on `_POLL_TIMEOUT`.

That should make it possible to reduce the initial job submission latency and reduce CPU usage when idling on most platforms at the cost of only 2 additional fds.

I am not sure if waiting on a pipe is possible and efficient under Windows though.

Reply:
BTW, I am fine with the name `_ThreadWakeup` personally.

Reply:
Here is what I meant about using a "waitable" pipe:

```python
>>> import os
>>> from multiprocessing.connection import wait
>>> r, w = os.pipe()
>>> fr = os.fdopen(r, mode='rb')
>>> fw = os.fdopen(w, mode='wb')
>>> wait([fr], timeout=1)
[]  # empty list after waiting for 1 second
>>> fw.write(b'\x00')
1
>>> fw.flush()
>>> wait([fr], timeout=1)
[<_io.BufferedReader name=149>]   # no wait here: instant result
```

Reply:
@pitrou do you know if using a pipe for waking up the call to `wait` in the main loop of the queue management thread is guaranteed to work on all supported platforms?

Reply:
I don't think so. You have to use a dedicated primitive on Windows.

Reply:
What are you trying to do exactly?

Reply:
To avoid dealing with different primitive, we can use a `multiprocessing.Pipe`.  
But I think we ruled that out before because of performance issues (_eg_ submitting many small tasks).

@pitrou : the goal of the `wakeup` object is to separate the waking up of the `queue_management_thread` from the `result_queue` to ensure that we can still communicate with it if the queue is deadlocked. This object right now is a flag shared between the threads, and thus we cannot use it in the `wait` primitive with it.

Reply:
This is only require for communication between the main thread (or some other user-spawn thread) and the queue management thread. We don't really need interprocess communication primitive here but it would be nice to avoid doing the 1ms wait loop and only rely on `select` to find which event to process next: result in the queue, process died or executor terminated or new process to create because we are submitting a job and the process count need to be adjusted. The last 2 classes or event need this waitable "wakeup" communication channel between threads.

Reply:
Why don't you use a Queue or Event?

Reply:
`Event` is not waitable via `multiprocessing.connection.wait` (at line https://github.com/python/cpython/pull/3895/files#diff-d24fedf7a1cf058e9e4166d89f2bb378R356). I was reluctant to use a `Queue` to avoid adding the overhead of an additional helper feeder thread. But actually, we could probably use a `SimpleQueue` whose `_reader` attribute should be waitable in a cross-platform manner (I think).

Reply:
Under the hood it's using a `multiprocess.connection.Pipe` but the `SimpleQueue` API is higher level.

Reply:
Note that `SimpleQueue` is a `Pipe` protected by a `Lock`. To wakeup the queue management thread when submitting a new task or when shutting down the executor, we really need to just send an arbitrary marker such as a single byte and flush. In master we are sending a None object via the result queue at the moment. So there is no need for a lock. We could therefore slightly reduce the overhead in `submit` by directly using a `Pipe` instead of pickling `None` + lock acquire and release.

Reply:
I meant a simple threading Queue, not a heavy-weight multiprocessing object.

A thought: how about a separate thread that monitors processes using their sentinels and then signals the main management thread using a threading Queue?

Reply:
The usage ot a `queue.Queue` does not really present advantages over the actual implementation which relies on the current `_Sentinel` object, as we only use it as a notifier, we do not care if there is multiple messages. I think that we fixed the `_POLL_TIMEOUT` such it has the same precision as the `wait` function so the overhead should be minim compared to the usage of a `Pipe` but it is a bit messier.
The benefit of using a `multiprocessing.Pipe` is that we can wait on it with `select`. This way, there is a single call responsible for the wait on the `_result_queue`, the sentinels and the `thread_wakeup`.

The experience of adding an extra thread to manage the processes has been done with `multiprocessing.Pool`. I do not remember perfectly why but it seems to me that it lead to extra complexity in the code. But I can look into it if you feel it is a better design for the executor.

Reply:
> I meant a simple threading Queue, not a heavy-weight multiprocessing object.

But then it's not possible to use a call to `wait` to wait at the same time for the next submit, for the next result or a worker process to die as we would. The current solution in this PR is introducing a polling mechanism with `_POLL_TIMEOUT` and I was thinking we would rather no use that to avoid CPU usage when polling on a idling executor and latency when submitting new tasks.

> A thought: how about a separate thread that monitors processes using their sentinels and then signals the main management thread using a threading Queue?

I think the code can stay simpler (easier to debug and maintain) by not introducing yet a new thread.

Reply:
I think we removed most (all?) sources of polling in multiprocessing some time ago. I don't want to reintroduce any; spurious wakeups can be very bad for power consumption on mobile devices (think laptops). I'll let you find the best solution to avoid them :-)